src/libfsm/reverse.c

```diff
@@ -47,86 +42,72 @@ fsm_reverse(struct fsm *fsm)
 	 */
 	endset = NULL;
 
-	/*
-	 * Create states corresponding to the original FSM's states.
-	 * These are created in the same order, due to fsm_addstate()
-	 * placing them at fsm->tail.
+	/* Perform the actual reversal by listing the reversed edges we need,
+	 * destroying all current edges, then adding those.
 	 */
-	/* TODO: possibly centralise as a state-copying function */
 	{
 		struct fsm_state *s;
+		struct set_iter it;
 
-		for (s = fsm->sl; s != NULL; s = s->next) {
-			struct fsm_state *p;
-
-			p = fsm_addstate(new);
-			if (p == NULL) {
-				fsm_free(new);
-				return 0;
-			}
-
-			s->equiv = p;
+		struct {
+			struct fsm_state *from;
+			struct fsm_state *to;
+			enum fsm_edge_type symbol;
+		} *q = NULL;
+		size_t lq = 0, nq = 0;
+		size_t i;
 
-			if (s == fsm->start) {
-				end = p;
-				fsm_setend(new, end, 1);
+		for (s = fsm->sl; s != NULL; s = s->next) {
+			struct fsm_edge *e;
+			for (e = set_first(s->edges, &it); e != NULL; e = set_next(&it)) {
+				struct fsm_state *to;
+				struct set_iter jt;
+				for (to = set_first(e->sl, &jt); to != NULL; to = set_next(&jt)) {
+					if (lq <= nq) {
+						void *tmp;
+						lq = lq > 0 ? lq * 2 : 1024;
+						tmp = realloc(q, lq * sizeof *q);
+						if (!tmp) {
+							free(q);
+							set_free(endset);
+							return 0;
+						}
+					}
+					q[nq].from = s;
+					q[nq].to = to;
+					q[nq].symbol = e->symbol;
+					nq++;
+				}
+				set_free(e->sl);
+				free(e);
 			}
+			set_clear(s->edges);
 
 			if (fsm_isend(fsm, s)) {
 				if (!set_add(&endset, s)) {
 					set_free(endset);
-					fsm_free(new);
 					return 0;
 				}
 
 				if (fsm->endcount == 1) {
-					assert(new->start == NULL);
-					new->start = p;
+					fsm->start = s;
 				}
 			}
-		}
-	}
-
-	if (end != NULL && fsm->opt->carryopaque != NULL) {
-		fsm->opt->carryopaque(endset, new, end);
-	}
-
-	/* Create reversed edges */
-	{
-		struct fsm_state *s;
 
-		for (s = fsm->sl; s != NULL; s = s->next) {
-			struct fsm_state *to;
-			struct fsm_state *se;
-			struct set_iter it;
-			struct fsm_edge *e;
-
-			to = s->equiv;
-
-			assert(to != NULL);
-
-			for (e = set_first(s->edges, &it); e != NULL; e = set_next(&it)) {
-				struct set_iter jt;
-
-				for (se = set_first(e->sl, &jt); se != NULL; se = set_next(&jt)) {
-					struct fsm_state *from;
-					struct fsm_edge *edge;
-
-					assert(se != NULL);
+			/* deliberately don't go via the API. we'll correct the end count when we're finished with it. */
```

Comment:
I don't want to change fsm->endcount. Probably better to save the original count instead.